napari/plugins/manager.py

```diff
@@ -0,0 +1,111 @@
+import importlib
+import os
+import pkgutil
+from logging import Logger
+
+import pluggy
+
+from . import _builtins, hookspecs
+
+logger = Logger(__name__)
+
+
+class NapariPluginManager(pluggy.PluginManager):
+    PLUGIN_ENTRYPOINT = "napari.plugin"
+    PLUGIN_PREFIX = "napari_"
+
+    def __init__(self, autodiscover=True):
+        """pluggy.PluginManager subclass with napari-specific functionality
+
+        In addition to the pluggy functionality, this subclass adds
+        autodiscovery using package naming convention.
+
+        Parameters
+        ----------
+        autodiscover : bool, optional
+            Whether to autodiscover plugins by naming convention and setuptools
+            entry_points, by default True
+        """
+        super().__init__("napari")
+
+        # define hook specifications and validators
+        self.add_hookspecs(hookspecs)
+
+        # register our own built plugins
+        self.register(_builtins, name='builtins')
+        # discover external plugins
+        if not os.environ.get("NAPARI_DISABLE_PLUGIN_AUTOLOAD"):
+            if autodiscover:
+                self.discover()
```

Comment:
I'm kind of in favour of having autodiscovery be off by default and provide users an easy way to launch it, or turn it on with an environment variable, at least until we experiment and also have a comprehensive model about stability and security of plugins... What do you think?

Reply:
hmmm.  this may come back to bite me, but I'd like for it to be on by default (definitely, at least, until #834 gets some love and we have some persistence juice).  This comes from the original desire of being able to help people with `pip install napari napari-plugin`.  If we turn off discovery, then they will have to enable plugins every time they start the program.  I'd also hate to start seeing `plugin_manager.discover()` in scripts all over the place.  and I don't love an env var if the group we're targeting here is less code-savvy users.

It's definitely not without risks though, and I'm certain that someone will create/install a plugin that will break napari and we'll have to figure it out... but the current strategy to deal with that is this:
1) if their plugin simply doesn't import (like, it raises an import error), then it will get logged, dismissed, and unregistered (tested by `napari_unimportable_plugin`)
2) if their plugin imports but fails to validate against our hookspec, same thing... dismiss and move on to the next one. (tested by `napari_invalid_plugin`)
3) if their plugin imports and validates, but has some _functional_ failures that would cause napari errors, that is what #935 is for (and is tested tested with `napari_bad_plugin`)

so I was hoping it would be enough to be aggressive about dismissing misbehaving plugins with helpful feedback, rather than keeping them all opt-in at this point.  But that's a big decision and you're welcome to just dictate that one.

i see this as a big focus for follow-up PRs.  and want to hammer it from many perspectives.  relatedly, I want to confirm is the the pluggy `load_setuptools_entrypoints` is equally safe for our use.  and if not I will override it.

Reply:
after looking into it, I chose to override the pluggy `load_setuptools_entrypoints` so that we can do some more try/catching.  Their default is just to have it raise an exception... I'd rather catch that exception and disable the plugin, and move on to the next one.

I also added a bunch more fixtures to thorough test both "unimportable" as well as invalid plugins that are registered via entry points.  I think (hope) the worst case scenario at the moment is that a user is frustrated by not being able to use a misbehaving plugin, rather than it crashing napari.  but again... I welcome scrutiny on this one.

Reply:
I also like "on by default" with it becoming an option in #834. I think right now there are a lot of people that are excited to try this out for whom not on by default will be a pain, and I'd much rather find out if there are problems faster at this point when we can quickly and easily fix them rather than wait many months before this becomes more widespread and only discover them then

Reply:
Yep, happy with all these rationales and the exception handling. One thing I would say is, if a package has a development home page listed in its metadata, we could put it in the "plugin dismissed" warning, "please see {url} for how to report this failure".

Reply:
yeah, love that idea.  Definitely want to make it obvious to the user if we are nuking one of their plugins and what to do next.  I also think this would be a good case for #794